diff --git a/dist/src/index.js b/dist/src/index.js
index ef02fd0564234b67f3e193acd59ba58f7e251f93..0bd66c2a59add7fdc0ed7ec23ac7b73964028105 100644
--- a/dist/src/index.js
+++ b/dist/src/index.js
@@ -475,7 +475,7 @@ export class GossipSub extends TypedEventEmitter {
         // TODO make this behavior more robust
         // This behavior is different than for inbound streams
         // If an outbound stream already exists, don't create a new stream
-        if (this.streamsOutbound.has(id)) {
+        if (this.streamsOutbound.get(id)?.rawStream?.status === "open") {
             return;
         }
         try {
@@ -1320,7 +1320,7 @@ export class GossipSub extends TypedEventEmitter {
     async directConnect() {
         const toconnect = [];
         this.direct.forEach((id) => {
-            if (!this.streamsOutbound.has(id)) {
+            if (!this.streamsOutbound.get(id)?.rawStream?.status !== "open") {
                 toconnect.push(id);
             }
         });
