diff --git a/dist/libp2p/config.js b/dist/libp2p/config.js
index ca49044ac91267e18d3080014365092bf99e2a53..9294ed2d138166929004faaf0123d15ff644ac1a 100644
--- a/dist/libp2p/config.js
+++ b/dist/libp2p/config.js
@@ -27,9 +27,16 @@ import { gossipsub } from "@libp2p/gossipsub";
 import { noise } from "@chainsafe/libp2p-noise";
 import { yamux } from "@chainsafe/libp2p-yamux";
 import { circuitRelayTransport } from "@libp2p/circuit-relay-v2";
-import { identify } from "@libp2p/identify";
+import { identify, identifyPush } from "@libp2p/identify";
 import { webRTC } from "@libp2p/webrtc";
+import { ping } from "@libp2p/ping";
 import { webSockets } from "@libp2p/websockets";
+const optionsIdentify = {
+    maxMessageSize: 1e6,
+    maxInboundStreams: 50,
+    maxOutboundStreams: 50,
+  };
+  
 export class ServiceClefPrivée {
     privateKey;
     constructor(components) {
@@ -54,6 +61,7 @@ export const OptionsDéfautLibp2pNode = () => {
             denyDialMultiaddr: () => false,
         },
         services: {
+            ping: ping(),
             identify: identify(),
             pubsub: gossipsub({ allowPublishToZeroTopicPeers: true }), // Erreur de type dans @chainsafe/pubsub
             obtClefPrivée: (components) => new ServiceClefPrivée(components),
@@ -75,8 +83,10 @@ export const OptionsDéfautLibp2pNavigateur = () => {
             denyDialMultiaddr: () => false,
         },
         services: {
-            identify: identify(),
-            pubsub: gossipsub({ allowPublishToZeroTopicPeers: true }), // Erreur de type dans @chainsafe/pubsub,
+            ping: ping(),
+            identify: identify(optionsIdentify),
+            identifyPush: identifyPush(optionsIdentify),
+            pubsub: gossipsub({ allowPublishToZeroTopicPeers: true, runOnLimitedConnection: true, }), // Erreur de type dans @chainsafe/pubsub,
             obtClefPrivée: (components) => new ServiceClefPrivée(components),
         },
     };
diff --git a/dist/libp2p/utils.js b/dist/libp2p/utils.js
index 426a3ddaaf0009684feb3590484531c3033e7a02..b122f6f66436fce7568ff7d857dcded6abf13222 100644
--- a/dist/libp2p/utils.js
+++ b/dist/libp2p/utils.js
@@ -20,7 +20,21 @@ export const connecterPairs = async (libp2p1, libp2p2, options) => {
             const interval = setInterval(testConnecté, 100);
             testConnecté();
         });
-        await libp2p2.dial(adresse1);
+        libp2p2.dial(adresse1);
+        await new Promise((resolve) => {
+            const testConnecté = () => {
+                const adresse = libp2p1
+                    .getConnections()
+                    .filter((c) => c.remotePeer.toString() === libp2p2.peerId.toString())
+                    .pop();
+                if (adresse != null) {
+                    clearInterval(interval);
+                    resolve(adresse);
+                }
+            };
+            const interval = setInterval(testConnecté, 100);
+            testConnecté();
+        });
     }
     else {
         await libp2p2.peerStore.save(libp2p1.peerId, {
