diff --git a/dist/nested.js b/dist/nested.js
index 11488299e171784b2d9bc4a2b29aa9d387a9cdcf..2b8c21c4c160765dd1d337123f220d4faff80e68 100644
--- a/dist/nested.js
+++ b/dist/nested.js
@@ -1,17 +1,18 @@
 import { Ajv } from "ajv";
 import { joinKey, splitKey, asJoinedKey, } from "@orbitdb/nested-db";
 import { getJoinedKey, removeUndefinedProperties, } from "./utils.js";
+
 const nestedCache = new WeakMap();
 export const typedNested = ({ db, schema, }) => {
-    let dbCache = nestedCache.get(db);
+    let dbCache = nestedCache.get(db.address);
     if (dbCache) {
-        const cached = dbCache.get(schema);
+        const cached = dbCache.get(JSON.stringify(schema));
         if (cached)
             return cached;
     }
     else {
-        nestedCache.set(db, new WeakMap());
-        dbCache = nestedCache.get(db);
+        nestedCache.set(db.address, new WeakMap());
+        dbCache = nestedCache.get(db.address);
     }
     const ajv = new Ajv({ allowUnionTypes: true });
     const rootValidator = ajv.compile(schema);
