diff --git a/dist/src/gossipsub.js b/dist/src/gossipsub.js
index eecaa1f10d492630d5cab1c8bedc0aaa7ec18635..3ac7f14e5982651315e94951ee176855f55f7b06 100644
--- a/dist/src/gossipsub.js
+++ b/dist/src/gossipsub.js
@@ -477,7 +477,7 @@ export class GossipSub extends TypedEventEmitter {
         // TODO make this behavior more robust
         // This behavior is different than for inbound streams
         // If an outbound stream already exists, don't create a new stream
-        if (this.streamsOutbound.has(id)) {
+        if (this.streamsOutbound.get(id)?.rawStream?.status === "open") {
             return;
         }
         try {
@@ -1329,7 +1329,7 @@ export class GossipSub extends TypedEventEmitter {
     async directConnect() {
         const toconnect = [];
         this.direct.forEach((id) => {
-            if (!this.streamsOutbound.has(id)) {
+            if (!this.streamsOutbound.get(id)?.rawStream?.status !== "open") {
                 toconnect.push(id);
             }
         });
