diff --git a/dist/src/gossipsub.js b/dist/src/gossipsub.js
index bc1ffa34043269d17ea67f3f5695c88d8e9c093f..1734a6ae96e3cf01c9194ef4e51cc57c1565542a 100644
--- a/dist/src/gossipsub.js
+++ b/dist/src/gossipsub.js
@@ -478,7 +478,7 @@ export class GossipSub extends TypedEventEmitter {
         // TODO make this behavior more robust
         // This behavior is different than for inbound streams
         // If an outbound stream already exists, don't create a new stream
-        if (this.streamsOutbound.has(id)) {
+        if (this.streamsOutbound.get(id)?.rawStream?.status === "open") {
             return;
         }
         try {
@@ -1336,7 +1336,7 @@ export class GossipSub extends TypedEventEmitter {
     async directConnect() {
         const toconnect = [];
         this.direct.forEach((id) => {
-            if (!this.streamsOutbound.has(id)) {
+            if (!(this.streamsOutbound.get(id)?.rawStream?.status === "open")) {
                 toconnect.push(id);
             }
         });
@@ -1865,7 +1865,7 @@ export class GossipSub extends TypedEventEmitter {
         }
         const rpcBytes = RPC.encode(rpc);
         try {
-            outboundStream.push(rpcBytes);
+            outboundStream.push(rpcBytes).catch((e) => { if ((e).name !== "StreamStateError") throw e });
         }
         catch (e) {
             this.log.error(`Cannot send rpc to ${id}`, e);
