diff --git a/dist/index.js b/dist/index.js
index f51d4a2f0e821c94f78667305e8a618a4f4973f6..3bcce39076965307c68547ceda756ca395a077ad 100644
--- a/dist/index.js
+++ b/dist/index.js
@@ -194,7 +194,6 @@ var decrypt2 = (cipher, password) => {
 var symmetric = { encryptBytes: encryptBytes2, decryptBytes: decryptBytes2, encrypt: encrypt2, decrypt: decrypt2 };
 
 // src/index.ts
-await sodium7.ready;
 export {
   asymmetric,
   base58,
