diff --git a/src/index.d.ts b/src/index.d.ts
index 10ed1d3fb214b30abf121353fcb55c73be47c192..a4f19066c3aac65e7aa87bc6baa53e8bade3d1ed 100644
--- a/src/index.d.ts
+++ b/src/index.d.ts
@@ -37,6 +37,7 @@ declare module "@orbitdb/core" {
     entryStorage: Storage;
     indexStorage: Storage;
     referencesCount: number;
+    signal?: AbortSignal;
   }>;
 
   type CreateDatabaseOptions<T extends ServiceMap = ServiceMap> = {
@@ -256,6 +257,7 @@ declare module "@orbitdb/core" {
     orbitdb: OrbitDB;
     identities: IdentitiesType;
     address?: string;
+    signal?: AbortSignal;
   }): Promise<AccessController>;
 
   export type AccessController = {
@@ -306,7 +308,7 @@ declare module "@orbitdb/core" {
     ipfs?: Helia<Libp2p<T>>;
   }): Promise<{
     createIdentity: (options: object) => Promise<Identity>;
-    getIdentity: (hash: string) => Promise<Identity>;
+    getIdentity: (hash: string, signal?: AbortSignal) => Promise<Identity>;
     verifyIdentity: (identity: Identity) => Promise<boolean>;
     sign: (identity: Identity, data: string) => Promise<string>;
     verify: (
@@ -338,8 +340,8 @@ declare module "@orbitdb/core" {
   }
 
   export type Storage = {
-    put: (hash: string, data: unknown) => Promise<void>; // Todo: check if DagCborEncodable is appropriate here
-    get: (hash: string) => Promise<ByteView<unknown>>;
+    put: (hash: string, data: unknown, signal?: AbortSignal) => Promise<void>; // Todo: check if DagCborEncodable is appropriate here
+    get: (hash: string, signal?: AbortSignal) => Promise<ByteView<unknown>>;
   };
   export function IPFSBlockStorage<T extends ServiceMap = ServiceMap>(args: {
     ipfs: Helia<Libp2p<T>>;
