diff --git a/dist/set.js b/dist/set.js
index 826dab5cef0ca21ead3674477e41882a4270a05d..109d8564be3c31f928fca25039bb14d146710e40 100644
--- a/dist/set.js
+++ b/dist/set.js
@@ -1,6 +1,6 @@
 import { Database } from "@orbitdb/core";
 const type = "set";
-const SetDb = () => async ({ ipfs, identity, address, name, access, directory, meta, headsStorage, entryStorage, indexStorage, referencesCount, syncAutomatically, onUpdate, }) => {
+const SetDb = () => async ({ ipfs, identity, address, name, access, directory, meta, headsStorage, entryStorage, indexStorage, referencesCount, syncAutomatically, onUpdate, signal, }) => {
     const database = await Database({
         ipfs,
         identity,
@@ -15,6 +15,7 @@ const SetDb = () => async ({ ipfs, identity, address, name, access, directory, m
         referencesCount,
         syncAutomatically,
         onUpdate,
+        signal,
     });
     const { add, del, iterator, all } = SetApi({ database });
     return {
