diff --git a/dist/feed.js b/dist/feed.js
index da01415b4d4d2a50f062d3c2dd746fec0e0b8922..8b2a01cb26ec2e9ed98b3baa92c40aea69d83551 100644
--- a/dist/feed.js
+++ b/dist/feed.js
@@ -1,6 +1,6 @@
 import { Database } from "@orbitdb/core";
 const type = "feed";
-const Feed = () => async ({ ipfs, identity, address, name, access, directory, meta, headsStorage, entryStorage, indexStorage, referencesCount, syncAutomatically, onUpdate, }) => {
+const Feed = () => async ({ ipfs, identity, address, name, access, directory, meta, headsStorage, entryStorage, indexStorage, referencesCount, syncAutomatically, onUpdate, signal, }) => {
     const database = await Database({
         ipfs,
         identity,
@@ -15,6 +15,7 @@ const Feed = () => async ({ ipfs, identity, address, name, access, directory, me
         referencesCount,
         syncAutomatically,
         onUpdate,
+        signal
     });
     const { add, remove, iterator, all } = FeedApi({ database });
     return {
